### PR TITLE
Add potential rel/target functions.

### PR DESCRIPTION
**Pull request by Camwyn**
For discussion
Refs: 
https://tribe.slack.com/archives/C0258NY1C/p1591808143434300 (archived)
https://github.com/the-events-calendar/products-engineering/pull/34
https://developer.wordpress.org/reference/functions/wp_targeted_link_rel/
https://developer.wordpress.org/reference/functions/wp_targeted_link_rel_callback/

### CODE REVIEW

File: src/functions/template-tags/html.php [hunk 1]
```diff
@@ -123,8 +123,72 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+ * @since TBD
+ *
+ * @param string  $target   Value for the target attribute.
+ * @param array|string $rel Value(s) for rel attribute.
+ * @param boolean $echo     Whether to echo the string or return it.
+ *
+ * @return string The rel attribute.
+ */
+function tribe_target( $target, $rel = [], $echo = true ) {
+	if ( empty( $rel ) ) {
```
Comment (paulmskim) [inline src/functions/template-tags/html.php:169]: why are we returning early if `$rel` is empty?

File: src/functions/template-tags/html.php [hunk 2]
```diff
@@ -123,8 +123,72 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+ * Adds rel attribute for links.
+ *
+ * @since TBD
+ *
+ * @param string|array $rel  Value(s) for rel attribute.
+ * @param boolean      $echo Whether to echo the string or return it.
+ *
+ * @return string the rel attribute
+ */
+function tribe_rel( $rel, $echo = true ) {
```
Comment (paulmskim) [inline src/functions/template-tags/html.php:142]: maybe set a default for `$rel`?

File: src/functions/template-tags/html.php [hunk 3]
```diff
@@ -123,8 +123,72 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+	if ( empty( $rel ) ) {
+		return;
+	}
+
+	$attr = 'rel="' . esc_attr( $rel ) . '"';
```
Comment (paulmskim) [inline src/functions/template-tags/html.php:147]: you're going to need a check for if `$rel` is an array or string here.

File: src/functions/template-tags/html.php [hunk 4]
```diff
@@ -123,8 +123,72 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+
+	$attr = 'target="' . esc_attr( $target ) . '"';
+
+	if ( '_blank' === $target ) {
+		$rel[] = 'noopener';
+		$rel[] = 'nofollow';
+	}
+
+	if ( ! empty( $rel ) ) {
+		$attr .= tribe_rel( $rel, false );
```
Comment (paulmskim) [inline src/functions/template-tags/html.php:185]: Here, you'd get `'target="_blank"rel="array array"'`. or whatever the a stringified array outputs to be. Maybe add each attribute string to an `$attr` array and implode at the end.

File: src/functions/template-tags/html.php (hunk 1 above)
Comment (Camwyn) [reply src/functions/template-tags/html.php:169]: Should have been `$target`

File: src/functions/template-tags/html.php (hunk 2 above)
Comment (Camwyn) [reply src/functions/template-tags/html.php:142]: There is no default rel value - so if it's not set we don't add the attribute.

File: src/functions/template-tags/html.php [hunk 5]
```diff
@@ -123,8 +123,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+
+/**
+ * Adds rel attribute for links.
+ *
+ * @since TBD
+ *
+ * @param string|array $rel  Value(s) for rel attribute.
+ * @param boolean      $echo Whether to echo the string or return it.
+ *
+ * @return string the rel attribute
```
Comment (Camwyn) [inline src/functions/template-tags/html.php:140]:
```suggestion
 * @return string The rel attribute.
```

File: src/functions/template-tags/html.php [hunk 6]
```diff
@@ -123,8 +123,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+ * Adds target attribute for links.
+ * If target is '_blank' also adds appropriate rel attributes.
+ *
+ * @since TBD
+ *
+ * @param string  $target   Value for the target attribute.
+ * @param array|string $rel Value(s) for rel attribute.
+ * @param boolean $echo     Whether to echo the string or return it.
+ *
+ * @return string The rel attribute.
```
Comment (Camwyn) [inline src/functions/template-tags/html.php:170]:
```suggestion
 * @return string The target attribute. Potentially also the rel attribute.
```

File: src/functions/template-tags/html.php [hunk 7]
```diff
@@ -123,8 +123,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+
+/**
+ * Adds target attribute for links.
+ * If target is '_blank' also adds appropriate rel attributes.
+ *
+ * @since TBD
+ *
+ * @param string  $target   Value for the target attribute.
+ * @param array|string $rel Value(s) for rel attribute.
+ * @param boolean $echo     Whether to echo the string or return it.
```
Comment (sc0ttkclark) [inline src/functions/template-tags/html.php:168]:
```suggestion
 * @param string       $target Value for the target attribute.
 * @param array|string $rel    Value(s) for rel attribute.
 * @param boolean      $echo   Whether to echo the string or return it.
```

**Comment by Camwyn** [comment]
Keeping this ope for now, but https://github.com/moderntribe/tribe-common/pull/1417 probably would make this obsolete.

**Comment by cliffordp** [comment]
@Camwyn
A) I prefer not having $echo as a function arg, since you can just put 'echo' in front of the function's usage.
B) Maybe we just use [wp_targeted_link_rel()](https://developer.wordpress.org/reference/functions/wp_targeted_link_rel/) instead of (or within) our own function?

**Review by sc0ttkclark — DISMISSED** [review]

File: src/functions/template-tags/html.php [hunk 8]
```diff
@@ -79,9 +79,9 @@ function tribe_required( $required, $echo = true ) {
 
 		if ( ! $echo ) {
 			return $required_string;
-		} else {
-			echo $required_string;
 		}
+
+		echo $required_string;
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:84]: * :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed
* :warning: Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$required_string'
* :no_entry_sign: All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found '$required_string'.

<sub>hash: ef2739b5a7ce84b598c4be8a2321c466</sub>

File: src/functions/template-tags/html.php [hunk 9]
```diff
@@ -103,9 +103,9 @@ function tribe_required_label( $required, $echo = true ) {
...
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:108]: * :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed
* :warning: Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$required_string'
* :no_entry_sign: All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found '$required_string'.

<sub>hash: b63336377fd3de9df18d9e070f668a9e</sub>

File: src/functions/template-tags/html.php [hunk 10]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
 
 		if ( ! $echo ) {
 			return $disabled_string;
-		} else {
-			echo $disabled_string;
 		}
+
+		echo $disabled_string;
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:132]: * :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed
* :warning: Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$disabled_string'
* :no_entry_sign: All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found '$disabled_string'.

<sub>hash: 69a7b06973d6958b8891cf84689296b9</sub>

File: src/functions/template-tags/html.php [hunk 11]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+ * Adds rel attribute for links.
+ *
+ * @since TBD
+ *
+ * @param string|array $rel  Value(s) for rel attribute.
+ * @param boolean      $echo Whether to echo the string or return it.
+ *
+ * @return string The rel attribute.
+ */
+function tribe_rel( $rel, $echo = true ) {
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:146]: * :no_entry_sign: Function name "tribe_rel" is not in camel caps format
* :no_entry_sign: TRUE, FALSE and NULL must be uppercase; expected "TRUE" but found "true"
* :no_entry_sign: Opening brace should be on a new line

<sub>hash: 944be642cb9ec96bd13e48b3bd719946</sub>

File: src/functions/template-tags/html.php [hunk 12]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+	if ( empty( $rel ) ) {
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:147]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: 7ee795f62e5e9c186140d6333d57001f</sub>

File: src/functions/template-tags/html.php [hunk 13]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+		return;
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:148]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: dfcbaad2e84415c93737f4d16e12d30f</sub>

File: src/functions/template-tags/html.php [hunk 14]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+	}
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:149]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: 9a1e5024d2c64c368a50936f2752b7fe</sub>

File: src/functions/template-tags/html.php [hunk 15]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+
+	if ( is_array( $rel )) {
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:151]: * :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: There must be a single space before the last parenthesis
* :no_entry_sign: No space before closing parenthesis is prohibited
* :no_entry_sign: No space before closing parenthesis is prohibited

<sub>hash: 3f6245db7df64be2ed32a1e68a778e2e</sub>

File: src/functions/template-tags/html.php [hunk 16]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+		$rel = implode( ' ', $rel );
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:152]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: 00c9e23c43cbe447478536db77e151ee</sub>

File: src/functions/template-tags/html.php [hunk 17]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+	}
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:153]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: c283f64341f894afd50a7706399c6867</sub>

File: src/functions/template-tags/html.php [hunk 18]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+
+	$attr = 'rel="' . esc_attr( $rel ) . '"';
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:155]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: 91b895b8fb1fcd904b6102ae72d3cb77</sub>

File: src/functions/template-tags/html.php [hunk 19]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+
+	if ( ! $echo ) {
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:157]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: f54c3dfc4ab936b41a58be6ee45fbd10</sub>

File: src/functions/template-tags/html.php [hunk 20]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+		return $attr;
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:158]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: 8e0e2fde911addea7c1a1480f37221a0</sub>

File: src/functions/template-tags/html.php [hunk 21]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+	}
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:159]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: ac0ad475cad87bd331ccad67a7e7a3b2</sub>

File: src/functions/template-tags/html.php [hunk 22]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+
+	echo $attr;
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:161]: * :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed
* :warning: Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$attr'
* :no_entry_sign: All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found '$attr'.

<sub>hash: ffb0078aa54f22a9a880ec27984aad43</sub>

File: src/functions/template-tags/html.php [hunk 23]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+ *
+ * @since TBD
+ *
+ * @param string       $target Value for the target attribute.
+ * @param array|string $rel    Value(s) for rel attribute.
+ * @param boolean      $echo   Whether to echo the string or return it.
+ *
+ * @return string The target attribute. Potentially also the rel attribute.
+ */
+function tribe_target( $target, $rel = [], $echo = true ) {
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:176]: * :no_entry_sign: Function name "tribe_target" is not in camel caps format
* :no_entry_sign: TRUE, FALSE and NULL must be uppercase; expected "TRUE" but found "true"
* :no_entry_sign: Opening brace should be on a new line

<sub>hash: f861880e41fdad49d11ba087ed336a45</sub>

File: src/functions/template-tags/html.php [hunk 24]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+	if ( empty( $target ) ) {
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:177]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: e39e1128d90b180984fc252d6f34c61c</sub>

File: src/functions/template-tags/html.php [hunk 25]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+		return;
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:178]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: 0c0145e1f26e7ac259e02057534c9d20</sub>

File: src/functions/template-tags/html.php [hunk 26]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+	}
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:179]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: 0b7c1fa7242df82b9f9888be7d867154</sub>

File: src/functions/template-tags/html.php [hunk 27]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+
+	// Format $rel as an array.
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:181]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: a4412f06b2a016f27da257f25563863a</sub>

File: src/functions/template-tags/html.php [hunk 28]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+	if ( ! is_array( $rel ) ) {
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:182]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: b415af396f4cbbee5406525dee83ce23</sub>

File: src/functions/template-tags/html.php [hunk 29]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+		$rel = explode( ' ', $rel );
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:183]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: 83f83ec726979e5dac1ae65161f94ab5</sub>

File: src/functions/template-tags/html.php [hunk 30]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
 	}
+
+	$attr = 'target="' . esc_attr( $target ) . '"';
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:186]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: 5fdfc90841e8c81073b8370238d2a471</sub>

File: src/functions/template-tags/html.php [hunk 31]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+
+	// If we're opening in a new window, let's be safe.
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:188]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: 3c81653af207f052e348cc18bb195bf8</sub>

File: src/functions/template-tags/html.php [hunk 32]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+	if ( '_blank' === $target ) {
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:189]: * :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed
* :no_entry_sign: Usage of Yoda conditions is not allowed; switch the expression order

<sub>hash: a17d918a5abbd2be5570a988868525a8</sub>

File: src/functions/template-tags/html.php [hunk 33]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+		$rel[] = 'noopener';
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:190]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: a620a802a384a41fd21337cdaa41d97c</sub>

File: src/functions/template-tags/html.php [hunk 34]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+		$rel[] = 'nofollow';
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:191]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: df14e57bd4d28cf1318113e765013d14</sub>

File: src/functions/template-tags/html.php [hunk 35]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+	}
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:192]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: 289b5c70e4e2827578aba08f9b8404ad</sub>

File: src/functions/template-tags/html.php [hunk 36]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+
+	// If we have $rel, add a rel attribute.
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:194]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: 44905defe3ae3e012773afaf84e79b4f</sub>

File: src/functions/template-tags/html.php [hunk 37]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+	if ( ! empty( $rel ) ) {
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:195]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: ac786818c171152eba7118f50841d23f</sub>

File: src/functions/template-tags/html.php [hunk 38]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+		$attr .= ' ' . tribe_rel( $rel, false );
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:196]: * :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed
* :no_entry_sign: TRUE, FALSE and NULL must be uppercase; expected "FALSE" but found "false"

<sub>hash: 8e9c166f2af4fa4e287e53fa986492fc</sub>

File: src/functions/template-tags/html.php [hunk 39]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+	}
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:197]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: f9219ac41c2f8a7204373c020029f192</sub>

File: src/functions/template-tags/html.php [hunk 40]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+
+
+	if ( ! $echo ) {
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:200]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: 56d3414468067fae4dbefa7f7560743b</sub>

File: src/functions/template-tags/html.php [hunk 41]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+		return $attr;
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:201]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: 616723cdf88f2064902805c570f08acb</sub>

File: src/functions/template-tags/html.php [hunk 42]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+	}
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:202]: :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed

<sub>hash: e90e9c78b2bab3f91070c5e856b5c154</sub>

File: src/functions/template-tags/html.php [hunk 43]
```diff
@@ -127,8 +127,79 @@ function tribe_disabled( $disabled, $echo = true ) {
...
+
+	echo $attr;
```
Comment (tr1b0t) [inline src/functions/template-tags/html.php:204]: * :no_entry_sign: Spaces must be used to indent lines; tabs are not allowed
* :warning: Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$attr'
* :no_entry_sign: All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found '$attr'.

<sub>hash: fd963bc68b65b2497b54e58bb4ec82bd</sub>

**Review by tr1b0t — COMMENTED** [review]
:no_entry_sign: 51 error(s)
:warning: 5 warning(s)

**Comment by borkweb** [comment]
This is stale and archived elsewhere. Closing this PR and deleting the branch.

**Pull request closed without merging**